p2p/transport/webrtc/udpmux/muxed_connection.go

```diff
@@ -0,0 +1,164 @@
+// data.
+type muxedConnection struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	pq     *packetQueue
+	// list of remote addresses associated with this connection.
+	// this is useful as a mapping from [address] -> ufrag
+	addresses []string
+	ufrag     string
+	mux       *udpMux
+}
+
+func newMuxedConnection(mux *udpMux, ufrag string) *muxedConnection {
+	ctx, cancel := context.WithCancel(mux.ctx)
+	return &muxedConnection{
+		ctx:    ctx,
+		cancel: cancel,
+		pq:     newPacketQueue(),
+		ufrag:  ufrag,
+		mux:    mux,
+	}
+}
+
+func (conn *muxedConnection) push(buf []byte, addr net.Addr) error {
+	return conn.pq.push(buf, addr)
+}
+
+// Close implements net.PacketConn
+func (conn *muxedConnection) Close() error {
+	_ = conn.closeConnection()
+	conn.mux.RemoveConnByUfrag(conn.ufrag)
+	return nil
+}
+
+// LocalAddr implements net.PacketConn
+func (conn *muxedConnection) LocalAddr() net.Addr {
+	return conn.mux.socket.LocalAddr()
+}
+
+// ReadFrom implements net.PacketConn
+func (conn *muxedConnection) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+	return conn.pq.pop(conn.ctx, p)
+}
+
+// SetDeadline implements net.PacketConn
+func (*muxedConnection) SetDeadline(t time.Time) error {
+	// no deadline is desired here
+	return nil
+}
+
+// SetReadDeadline implements net.PacketConn
+func (*muxedConnection) SetReadDeadline(t time.Time) error {
+	// no read deadline is desired here
+	return nil
+}
+
+// SetWriteDeadline implements net.PacketConn
+func (*muxedConnection) SetWriteDeadline(t time.Time) error {
+	// no write deadline is desired here
+	return nil
+}
+
+// WriteTo implements net.PacketConn
+func (conn *muxedConnection) WriteTo(p []byte, addr net.Addr) (n int, err error) {
+	return conn.mux.writeTo(p, addr)
+}
+
+func (conn *muxedConnection) closeConnection() error {
+	select {
+	case <-conn.ctx.Done():
+		return errAlreadyClosed
+	default:
+	}
+	conn.pq.close()
+	conn.cancel()
+	return nil
+}
+
+type packet struct {
+	addr net.Addr
+	buf  []byte
+}
+
+var (
+	errTooManyPackets = errors.New("too many packets in queue; dropping")
+)
+
+// just a convenience wrapper around a channel
+type packetQueue struct {
+	pkts   chan packet
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newPacketQueue() *packetQueue {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &packetQueue{
+		ctx:    ctx,
+		cancel: cancel,
+		pkts:   make(chan packet, maxPacketsInQueue),
+	}
+}
+
+// pop reads a packet from the packetQueue or blocks until
+// either a packet becomes available or the buffer is closed.
+func (pq *packetQueue) pop(ctx context.Context, buf []byte) (int, net.Addr, error) {
+	select {
+	case <-ctx.Done():
+		return 0, nil, ctx.Err()
+	// It is desired to allow reads of this channel even
+	// when pq.ctx.Done() is already closed.
+	case p, ok := <-pq.pkts:
+		if !ok {
+			return 0, nil, io.EOF
+		}
+		n := copy(buf, p.buf)
+		var err error
+		if n < len(p.buf) {
+			err = io.ErrShortBuffer
+		}
```

Comment:
Is this correct? Calling `ReadFrom` is valid even with small byte slices, but this would lead to errors and dropped packets here.

Reply:
I agree with your reasoning but that would need to refactor I think, going to give it some thought. Because right now the code would not be able to handle that given the buffer is already put back in the pool even on a short read.

Reply:
This will be easier if you have a slice of `[]byte` here instead of a chan (as I suggested in https://github.com/libp2p/go-libp2p/pull/1999/files/9545df6945926f2492c8e080986b8fc50394a8f9#r1103905493), since you then can just cut off bytes from the first `[]byte` if it's a short read.

Reply:
ok perfect. As it's already 1:35AM going to keep that one for tomorrow! Thanks for the suggestion, does look promising! Thank you a lot!

Reply:
Super, that now is a lot easier to implement indeed. Have done so as well.
Thanks for the tip!